File: demo-android/app/src/androidTest/java/com/example/demo/e2e/support/mock/WireMockClient.kt
package com.example.demo.e2e.support.mock

import com.example.demo.BuildConfig
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject

class WireMockClient(
    private val wiremockHost: String = resolveWiremockHost(),
) {
    private val client = OkHttpClient()

    fun clearMocks() {
        deleteAdminResource("/__admin/mappings")
    }

    fun resetScenarios() {
        postAdminResource("/__admin/scenarios/reset", "{}")
    }

    fun reset() {
        clearMocks()
        resetScenarios()
    }

    fun addScenarioMapping(
        scenarioName: String,
        request: JSONObject,
        response: JSONObject,
    ) {
        val requiredState = fetchScenarioState(scenarioName)
        val newState = nextScenarioState(requiredState)

        val mapping = JSONObject()
            .put("scenarioName", scenarioName)
            .put("requiredScenarioState", requiredState)
            .put("newScenarioState", newState)
            .put("request", request)
            .put("response", response)

        postAdminResource("/__admin/mappings", mapping.toString())
    }

    private fun fetchScenarioState(scenarioName: String): String {
        val response = client.newCall(
            Request.Builder()
                .url("$wiremockHost/__admin/scenarios")
                .build(),
        ).execute()
        check(response.isSuccessful) {
            "WireMock GET /__admin/scenarios failed: HTTP ${response.code} ${response.body?.string()}"
        }

        val scenarios = JSONObject(response.body!!.string()).getJSONArray("scenarios")
        for (index in 0 until scenarios.length()) {
            val scenario = scenarios.getJSONObject(index)
            if (scenario.getString("name") == scenarioName) {
                return scenario.getString("state")
            }
        }
        return "Started"
    }

    private fun nextScenarioState(currentState: String): String = when (currentState) {
        "Started" -> "step-1"
        else -> "step-${currentState.removePrefix("step-").toInt() + 1}"
    }

    private fun deleteAdminResource(path: String) {
        val response = client.newCall(
            Request.Builder()
                .delete()
                .url("$wiremockHost$path")
                .build(),
        ).execute()
        check(response.isSuccessful) {
            "WireMock DELETE $path failed: HTTP ${response.code} ${response.body?.string()}"
        }
    }

    private fun postAdminResource(path: String, body: String) {
        val response = client.newCall(
            Request.Builder()
                .url("$wiremockHost$path")
                .post(body.toRequestBody("application/json".toMediaType()))
                .build(),
        ).execute()
        check(response.isSuccessful) {
            "WireMock POST $path failed: HTTP ${response.code} ${response.body?.string()}"
        }
    }

    companion object {
        fun resolveWiremockHost(): String = BuildConfig.WIREMOCK_URL.trimEnd('/')
    }
}

File: e2e/maestro/support/mocks/WireMockClient.ts
import { testConfig } from '@/test.config';

type ScenarioMapping = {
  scenarioName: string;
  request: Record<string, string>;
  response: Record<string, unknown>;
};

const WIREMOCK_RETRY_ATTEMPTS = 5;
const WIREMOCK_RETRY_DELAY_MS = 200;

export class WireMockClient {
  private readonly baseUrl = testConfig.wiremock.url.replace(/\/$/, '');
  private readonly plannedScenarioStates = new Map<string, string>();

  async ensureReady(): Promise<this> {
    await this.requestAdmin('GET', '/__admin/health');
    return this;
  }

  async clearMocks(): Promise<this> {
    await this.requestAdmin('DELETE', '/__admin/mappings');
    return this;
  }

  async resetScenarios(): Promise<this> {
    await this.requestAdmin('POST', '/__admin/scenarios/reset', '{}');
    this.plannedScenarioStates.clear();
    return this;
  }

  async reset(): Promise<this> {
    await this.clearMocks();
    await this.resetScenarios();
    return this;
  }

  async addScenarioMapping(mapping: ScenarioMapping): Promise<this> {
    const requiredState =
      this.plannedScenarioStates.get(mapping.scenarioName) ?? 'Started';
    const newState = nextScenarioState(requiredState);
    this.plannedScenarioStates.set(mapping.scenarioName, newState);

    await this.requestAdmin(
      'POST',
      '/__admin/mappings',
      JSON.stringify({
        scenarioName: mapping.scenarioName,
        requiredScenarioState: requiredState,
        newScenarioState: newState,
        request: mapping.request,
        response: mapping.response,
      }),
    );

    return this;
  }

  private async requestAdmin(
    method: 'GET' | 'POST' | 'DELETE',
    path: string,
    body?: string,
  ): Promise<void> {
    const url = `${this.baseUrl}${path}`;
    let lastError: unknown;

    for (let attempt = 1; attempt <= WIREMOCK_RETRY_ATTEMPTS; attempt++) {
      try {
        const response = await fetch(url, {
          method,
          headers: body ? { 'Content-Type': 'application/json' } : undefined,
          body,
        });

        if (response.ok) {
          return;
        }

        lastError = new Error(`HTTP ${response.status}`);
      } catch (error) {
        lastError = error;
      }

      if (attempt < WIREMOCK_RETRY_ATTEMPTS) {
        await sleep(WIREMOCK_RETRY_DELAY_MS * attempt);
      }
    }

    const causeMessage =
      lastError instanceof Error ? lastError.message : String(lastError);

    throw new Error(
      `WireMock ${method} ${path} failed after ${WIREMOCK_RETRY_ATTEMPTS} attempts at ${this.baseUrl}. ` +
        `Is WireMock running? Start it with: docker compose -f docker/docker-compose/run-application.yml up -d qa-demo-wiremock. ` +
        `Cause: ${causeMessage}`,
      { cause: lastError },
    );
  }
}

function nextScenarioState(currentState: string): string {
  if (currentState === 'Started') {
    return 'step-1';
  }
  const step = Number.parseInt(currentState.replace('step-', ''), 10);
  return `step-${step + 1}`;
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}
